Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,390 @@
+/-
+Copyright (c) 2024 Enrico Z. Borba. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Enrico Z. Borba
+-/
+
+import Mathlib.Probability.Density
+import Mathlib.Probability.Notation
+import Mathlib.MeasureTheory.Constructions.Prod.Integral
+import Mathlib.Analysis.SpecialFunctions.Integrals
+import Mathlib.Probability.Distributions.Uniform
+
+/-!
+
+# Freek № 99: Buffon's Needle
+
+This file proves Theorem 99 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/), also
+known as Buffon's Needle, which gives the probability of a needle of length `l > 0` crossing any
+one of infinite vertical lines spaced out `d > 0` apart.
+
+The two cases are proven in `buffon_short` and `buffon_long`.
+
+## Overview of the Proof
+
+We define a random variable `B : Ω → ℝ × ℝ` with a uniform distribution on `[-d/2, d/2] × [0, π]`.
+This represents the needle's x-position and angle with respect to a vertical line. By symmetry, we
+need to consider only a single vertical line positioned at `x = 0`. A needle therefore crosses the
+vertical line if its projection onto the x-axis contains `0`.
+
+We define a random variable `N : Ω → ℝ` that is `1` if the needle crosses a vertical line, and `0`
+otherwise. This is defined as `fun ω => Set.indicator (needleProjX l (B ω).1 (B ω).2) 1 0`.
+f
+As in many references, the problem is split into two cases, `l ≤ d` (`buffon_short`), and `d ≤ l`
+(`buffon_long`). For both cases, we show that
+```
+ℙ[N] = (d * π) ⁻¹ *
+    ∫ θ in 0..π,
+      ∫ x in Set.Icc (-d / 2) (d / 2) ∩ Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2), 1
+```
```

Comment:
These will look nicer in the web docs as
````suggestion
```lean
ℙ[N] = (d * π) ⁻¹ *
    ∫ θ in 0..π,
      ∫ x in Set.Icc (-d / 2) (d / 2) ∩ Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2), 1
```
````